### ISSUE TITLE
Assembly.Location and Assembly.Codebase are empty when running with preloader

### ISSUE THREAD

**Issue by denikson** [BepInEx/BepInEx#55]
**Describe the bug**
When using the preloader to bootstrap BepInEx, `Assembly.Location` and `Assembly.CodeBase` are empty in preloaded assemblies.

**To Reproduce**
Create a basic plugin that has the following method:
```csharp
using UnityEngine;

void Awake() 
{
    Console.WrtieLine("Location of UnityEngine.dll");
    Console.WriteLine(typeof(Application).Assembly.Location);
    Console.WriteLine(typeof(Application).Assembly.CodeBase);
}
```
Run the game with preloader and console enabled (i.e. through UnityDoorstop).
It will show the following messages on game launch:
```
Location of UnityEngine.dll


```

**Expected behavior**
Assembly location and codebase should be set for all preloaded assemblies and the above example should print
```
Location of UnityEngine.dll
<path to UnityEngine.dll>
<URI to UnityEngine.dll>
```
In other words, the path to UnityEngine (and other assemblies in the `Managed` folder) should have their locations set correctly

**Desktop (please complete the following information):**
 - BepInEx version [e.g. 4.0]: 4.1.0.0

**Additional context**
The issue is caused by preloader loading assemblies from memory, in which case the CLR simply can't set the paths correctly. 
This should be fixable by caching assembly paths in a map during preloading and either patching `Assembly.Location` and `Assembly.CodeBase` via Harmony or by rerouting appropriate internal calls.
